_posts/2017-11-17-crates-are-not-safe.md

```diff
@@ -0,0 +1,100 @@
+major problems that are preventing us from using crates.io: `unsafe` and
+`std::`.
+
+## Crates are Mostly Unsafe
+
+Tock's kernel is designed to be split into two parts: a minimal but trusted core
+kernel and a much larger set of untrusted "capsules". The core kernel handles
+loading applications and controlling memory-mapped IO registers, and as such
+requires using `unsafe` code. The capsules, however, implement chip drivers and
+other utilities, and are strictly bound by the Rust typesystem (i.e. no
+`unsafe`). This divide allows us to maximize the safety guarantees of Rust while
+still making a kernel possible.
+
+This restriction on capsules has so far made it very difficult to use
+community-developed crates. Following up on @josh_triplett's [analysis of
+crates.io](https://twitter.com/josh_triplett/status/849089108727222275), we
+survey all of the crates on crates.io and find that 34% (4216/12360) of the
+crates use an `unsafe` block at least once. Of crates that have been downloaded
+10k times or more, 55% (522/944) use `unsafe`. While these ratios are not
+promising, it may be possible to work around the unsafe crates. The problem gets
+worse, however, when we consider dependencies. A crate that does not use
+`unsafe` but relies on a crate that does is effectively unsafe itself. After
+tracking required "normal" dependencies, 77% (9452/12360) of crates are unsafe
+(the number rises to 78% (9624/12360) if optional dependencies are included).
+This means that most of the 12,000-odd crates on crates.io are unsafe!
+
+In Tock we mark the capsules crate with `#![forbid(unsafe_code)]` to ensure that
+no `unsafe` code can creep in. To see if others are doing this as well, we also
+scan crates.io for the use of that flag. We find that only 0.2% (21/12360) of
+crates are explicitly disallowing unsafe code. Because there is no way to use
+cargo or the compiler to enforce that dependencies are safe (although we floated
+the [idea](https://internals.rust-lang.org/t/pre-rfc-cargo-safety-rails/5535)),
+we check to see if any of those explicitly-safe crates have unsafe dependencies.
+We find that all 21 crates have dependencies that use `unsafe`, effectively
+making their use of `#![forbid(unsafe_code)]` moot.
+
+Of course, it is entirely possible that all uses of `unsafe` in crates.io are
+valid, necessary, and actually safe. But validating this, particularly as crates
+are updated, for all dependencies would be a significant challenge. It seems
+that the Rust ecosystem needs a better mechanism for helping developers ensure
+that the code they are using is actually type-safe.
+
+## Usage of the Standard Library is Pervasive
+
+The other challenge for Tock, as an embedded operating system, is the prevalent
+use of the standard library on crates.io. Running on bare-metal, Tock cannot use
+`std::` and only has access to the libraries provided by `core::`. Any
+dependency we would want to use for Tock needs to include `#![no_std]` so that
```

Comment:
Maybe this post should, more generally, talk about what people writing capsules, in general (not just us "for Tock") can or can't use.